### ISSUE TITLE
nightmare.on('page') not fired for various kinds of console output

### ISSUE THREAD

**Issue by keithkml** [segmentio/nightmare#1010]
I'm trying to diagnose an error on our continuous build server, and seeing the Electron console would really help. It appears that Nightmare allows listening for these events, but unfortunately they don't seem to be firing for any of the errors I'm seeing.

I'd expect to see the 'BROWSER EVENT' line in my terminal a few times, one line each for those HTTP, websocket, and source map errors in the screen shot below. However in fact I'm not seeing the `page` event ever being fired.

```js
function startBrowser() {
  return Nightmare({show: true, dock: true, openDevTools: {mode: 'detach'}})
    .onBeforeSendHeaders((details, callback) => {
      callback({
        cancel: false, requestHeaders: {
          'xxx': 'yyy'
        }
      });
    });
}

test('search should work', assert => {
  const browser = startBrowser();
    function saveScreenshotAndEnd(savePath) {
      return browser.screenshot().end().then(imageBuffer => {
        const result = fs.writeFileSync(savePath, imageBuffer);
        shutdown();
        return result;
      });
    }
    browser
      /* eslint-disable no-console */
      .on('console', (logType, ...args) => console[logType]('BROWSER CONSOLE:', ...args))
      .on('page', (type, message, stack) => console.log('BROWSER EVENT:', type, message, stack))
      /* eslint-enable no-console */

      .goto('http://localhost:15872/search')

      // Search for the special e2e test account
      // .wait('.searchInput')
      .type('.searchInput', 'KeithTestNEVERDELETE')

      // Click the first result
      .wait('.searchResults a[href]')
      .click('.searchResults a[href]')
      .wait('h1')

      // Check the text
      .evaluate(() => document.querySelector('h1').innerText) // eslint-disable-line no-undef
      .then(headline => assert.equal(headline, 'KeithTestNEVERDELETE'))

      // This .then() is called after the whole Nightmare session is over
      .then(() => {
        saveScreenshotAndEnd('artifacts/success.png');
      })

      // And this is called if any of those steps fail
      .catch(error => {
        // eslint-disable-next-line no-console
        console.error(error);
        assert.error(error);
        saveScreenshotAndEnd('artifacts/failure.png');
      });
  });
});
```

<img width="468" alt="screen shot 2017-02-22 at 12 48 31 pm" src="https://cloud.githubusercontent.com/assets/7430512/23224784/aa6e6a20-f8fd-11e6-9cf4-8072277cefdc.png">

**Comment by rosshinkley** [comment]
Hm, I _think_ the `page` event - specifically with errors - is for on-page javascript errors (I can't remember offhand if it's parsing and uncaught exceptions or just uncaught exceptions).  I don't think it will capture sourcemap parsing issues or websocket connection issues.

For the `WebSocket`, I'm not sure you'll be able to capture the error short of writing a wrapper you can load up with preload, kind of akin to how `console` et al is handled.  Nothing jumps out at me in the Electron docs.  (Someone else out there happen to know?)  The sourcemap errors seem like they _should_ cause something to get called, but again, nothing jumps out at me.

Pondering out loud, have you given the Electron [`debugger`](http://electron.atom.io/docs/api/debugger/) a look?  You might be able to use `debugger.on('message')` to capture those problem events using a custom action.

**Comment by keithkml** [comment]
Wow thank you for doing so much research. I'll look into debugger. Feel free to close this since it sounds like it's not an issue with nightmare itself.

**Comment by rosshinkley** [comment]
@keithkml If you come up with a custom action using the debugger, I'd be _very_ interested in seeing it.

Also, I neglected to mention - if you go this route, you will likely not be able to use the debugger tray (or you'll have to be _very_ judicious about opening and closing it) as Chromium only allows one attachment to the debugger at any given time.  I don't know that you'd hit that, but I thought I'd say it out loud to prevent you from head+desk.

Closing for the moment.
